lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -0,0 +1,683 @@
+  }
+  enum {
+    NumLowBitsAvailable = PointerLikeTypeTraits<Attribute>::NumLowBitsAvailable
+  };
+};
+
+} // namespace llvm
+
+//===----------------------------------------------------------------------===//
+// Static information gathered once upfront
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// Attributes used throughout the annotations.
+struct Strings {
+  MLIRContext *const cx;
+  Strings(MLIRContext *cx) : cx(cx) {}
+
+  Identifier annos = Identifier::get("annotations", cx);
+  Identifier fannos = Identifier::get("firrtl.annotations", cx);
+
+  StringAttr dataTapsClass =
+      StringAttr::get(cx, "sifive.enterprise.grandcentral.DataTapsAnnotation");
+  StringAttr memTapClass =
+      StringAttr::get(cx, "sifive.enterprise.grandcentral.MemTapAnnotation");
+
+  StringAttr deletedKeyClass =
+      StringAttr::get(cx, "sifive.enterprise.grandcentral.DeletedDataTapKey");
+  StringAttr literalKeyClass =
+      StringAttr::get(cx, "sifive.enterprise.grandcentral.LiteralDataTapKey");
+  StringAttr referenceKeyClass =
+      StringAttr::get(cx, "sifive.enterprise.grandcentral.ReferenceDataTapKey");
+  StringAttr internalKeyClass = StringAttr::get(
+      cx, "sifive.enterprise.grandcentral.DataTapModuleSignalKey");
+};
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// Data Taps Implementation
+//===----------------------------------------------------------------------===//
+
+// static LogicalResult processDataTapAnnotation(FExtModuleOp module,
+//                                               DictionaryAttr anno) {}
+
+/// A port annotated with a data tap key or mem tap.
+struct AnnotatedPort {
+  unsigned portNum;
+  DictionaryAttr anno;
+};
+
+/// An extmodule that has annotated ports.
+struct AnnotatedExtModule {
+  FExtModuleOp extModule;
+  SmallVector<AnnotatedPort, 4> portAnnos;
+  ArrayAttr filteredModuleAnnos; // module annotations without data tap stuff
+};
+
+/// A value annotated to be tapped.
+struct TappedValue {
+  Value value;
+  DictionaryAttr anno;
+};
+
+/// An absolute instance path.
+using InstancePath = ArrayRef<InstanceOp>;
+
+template <typename T>
+T &operator<<(T &os, const InstancePath &path) {
+  os << "$root";
+  for (auto inst : path)
+    os << "." << inst.name();
+  // if (path.empty())
+  //   return os << "$root";
+  // llvm::interleave(
+  //     path.begin(), path.end(), [&](InstanceOp op) { os << op.name(); },
+  //     [&] { os << "."; });
+  return os;
+}
+
+/// A data structure that tracks the instances of modules and can provide
+/// absolute paths to these instances.
+struct InstanceGraph {
+  /// The root circuit.
+  CircuitOp circuitOp;
+  /// The main module in the circuit.
+  Operation *mainModule;
+  /// A mapping from a module to all its instances in the design.
+  DenseMap<Operation *, SmallPtrSet<InstanceOp, 1>> moduleInstances;
```

Comment:
That's an excellent point, and the set's deduplication effect is not needed either since the way the map is populated (walking the IR) will only ever hit an `InstanceOp` once. Also drops the need for the pointer traits above (but the TableGen change is probably still worth discussing).